File: src/MainListActivity.scala
package org.aprsdroid.app

import _root_.android.content.{BroadcastReceiver, Context, Intent, IntentFilter}
import _root_.android.graphics.PorterDuff
import _root_.android.view.View.OnClickListener
import _root_.android.view.{ContextMenu, Menu, MenuItem, View, Window}
import _root_.android.widget.Button
import _root_.androidx.core.content.ContextCompat

class MainListActivity(actname : String, menuid : Int) extends LoadingListActivity
		with OnClickListener {
	menu_id = menuid

	lazy val singleBtn = findViewById(R.id.singlebtn).asInstanceOf[Button]
	lazy val startstopBtn = findViewById(R.id.startstopbtn).asInstanceOf[Button]

	lazy val miclReceiver = new BroadcastReceiver() {
		override def onReceive(ctx : Context, i : Intent) {
			setProgress(i.getIntExtra("level", 100)*99)
		}
	}

	lazy val linkOnOffReceiver = new BroadcastReceiver() {
		override def onReceive(ctx : Context, i : Intent) {
			setTitleStatus()
		}
	}

	def onContentViewLoaded() {
		singleBtn.setOnClickListener(this);
		startstopBtn.setOnClickListener(this);
		registerForContextMenu(getListView())
	}

	override def onResume() {
		super.onResume()
		checkConfig()

		setTitleStatus()
		setupButtons(AprsService.running)
		makeLaunchActivity(actname)

		ContextCompat.registerReceiver(this, miclReceiver, new IntentFilter(AprsService.MICLEVEL), Context.RECEIVER_NOT_EXPORTED)
		ContextCompat.registerReceiver(this, linkOnOffReceiver, new IntentFilter(AprsService.SERVICE_STOPPED), Context.RECEIVER_EXPORTED)
		ContextCompat.registerReceiver(this, linkOnOffReceiver, new IntentFilter(AprsService.LINK_OFF), Context.RECEIVER_EXPORTED)
		ContextCompat.registerReceiver(this, linkOnOffReceiver, new IntentFilter(AprsService.LINK_ON), Context.RECEIVER_EXPORTED)
	}
	override def onPause() {
		super.onPause()
		try { unregisterReceiver(miclReceiver) } catch { case _ : Throwable => }
		try { unregisterReceiver(linkOnOffReceiver) } catch { case _ : Throwable => }
	}

	def setupButtons(running : Boolean) {
		//singleBtn.setEnabled(!running)
		if (running) {
			startstopBtn.getBackground().setColorFilter(0xffffc0c0, PorterDuff.Mode.MULTIPLY)
			startstopBtn.setText(R.string.stoplog)
		} else {
			startstopBtn.getBackground().setColorFilter(0xffc0ffc0, PorterDuff.Mode.MULTIPLY)
			startstopBtn.setText(R.string.startlog)
		}
	}

	override def onClick(view : View) {
		view.getId match {
		case R.id.singlebtn =>
			startAprsService(START_SERVICE_ONCE)
			setupButtons(true)
		case R.id.startstopbtn =>
			val is_running = AprsService.running
			if (!is_running) {
				startAprsService(START_SERVICE)
			} else {
				stopAprsService()
			}
			setupButtons(!is_running)
		}
	}

	override def onStopLoading() {
		super.onStopLoading()
		setupButtons(AprsService.running)
	}

}

File: src/MainListActivity.kt
package org.aprsdroid.app

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.PorterDuff
import android.view.View
import android.widget.Button
import androidx.core.content.ContextCompat

abstract class MainListActivity(
    val actname: String,
    menuid: Int
) : LoadingListActivity(), View.OnClickListener {

    init {
        menu_id = menuid
    }

    val singleBtn: Button by lazy { findViewById(R.id.singlebtn) }
    val startstopBtn: Button by lazy { findViewById(R.id.startstopbtn) }

    val miclReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context, i: Intent) {
            @Suppress("DEPRECATION")
            setProgress(i.getIntExtra("level", 100) * 99)
        }
    }

    val linkOnOffReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context, i: Intent) {
            setTitleStatus()
        }
    }

    fun onContentViewLoaded() {
        singleBtn.setOnClickListener(this)
        startstopBtn.setOnClickListener(this)
        registerForContextMenu(listView)
    }

    fun setTitleStatus() {
        title = if (AprsService.running) {
            getString(R.string.app_name) + " (" + prefs.getCallSsid() + ")"
        } else {
            getString(R.string.app_name)
        }
    }

    @SuppressLint("WrongConstant")
    override fun onResume() {
        super.onResume()
        setTitleStatus()
        setupButtons(AprsService.running)

        ContextCompat.registerReceiver(this, miclReceiver, IntentFilter(AprsService.MICLEVEL), ContextCompat.RECEIVER_NOT_EXPORTED)
        ContextCompat.registerReceiver(this, linkOnOffReceiver, IntentFilter(AprsService.SERVICE_STOPPED), ContextCompat.RECEIVER_EXPORTED)
        ContextCompat.registerReceiver(this, linkOnOffReceiver, IntentFilter(AprsService.LINK_OFF), ContextCompat.RECEIVER_EXPORTED)
        ContextCompat.registerReceiver(this, linkOnOffReceiver, IntentFilter(AprsService.LINK_ON), ContextCompat.RECEIVER_EXPORTED)
    }

    override fun onPause() {
        super.onPause()
        try { unregisterReceiver(miclReceiver) } catch (_: Throwable) {}
        try { unregisterReceiver(linkOnOffReceiver) } catch (_: Throwable) {}
    }

    fun setupButtons(running: Boolean) {
        if (running) {
            startstopBtn.background.setColorFilter(0xffffc0c0.toInt(), PorterDuff.Mode.MULTIPLY)
            startstopBtn.setText(R.string.stoplog)
        } else {
            startstopBtn.background.setColorFilter(0xffc0ffc0.toInt(), PorterDuff.Mode.MULTIPLY)
            startstopBtn.setText(R.string.startlog)
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.singlebtn -> {
                startService(AprsService.intent(this, AprsService.SERVICE_ONCE))
                setupButtons(true)
            }
            R.id.startstopbtn -> {
                val isRunning = AprsService.running
                if (!isRunning) {
                    startService(AprsService.intent(this, AprsService.SERVICE))
                } else {
                    startService(AprsService.intent(this, AprsService.SERVICE_STOP))
                }
                setupButtons(!isRunning)
            }
        }
    }

    override fun onStopLoading() {
        super.onStopLoading()
        setupButtons(AprsService.running)
    }
}
